backend/dataall/base/utils/expiration_util.py

```diff
@@ -8,29 +8,41 @@ class ExpirationUtils:
+            currentDate.year, currentDate.month
+        )
         if expirySetting == Expiration.Quartely.value:
-            if currentDate < datetime(currentDate.year, currentDate.month, 15).date():
-                # First half of the month - extend 2.X months
-                quarterlyCalculatedDate = currentDate + relativedelta(months=expirationPeriod * 3 - 1)
-            else:
-                # Second half of the month - extend 3.X months
+            if last_week_start_date <= currentDate <= last_week_end_date:
+                # If in the last week of the month extend x3 months
                 quarterlyCalculatedDate = currentDate + relativedelta(months=expirationPeriod * 3)
-            day = calendar.monthrange(quarterlyCalculatedDate.year, quarterlyCalculatedDate.month)[1]
-            shareExpiryDate = datetime(quarterlyCalculatedDate.year, quarterlyCalculatedDate.month, day)
-        elif expirySetting == Expiration.Monthly.value:
-            if currentDate < datetime(currentDate.year, currentDate.month, 15).date():
-                # First half of the month - extend until end of month
-                monthlyCalculatedDate = currentDate + relativedelta(months=expirationPeriod - 1)
             else:
-                # Second half of the month - extend until end of next month
+                # If the user is not in the last week, consider this month into the calculation
+                quarterlyCalculatedDate = currentDate + relativedelta(months=expirationPeriod * 3 - 1)
+            monthEndDay = calendar.monthrange(quarterlyCalculatedDate.year, quarterlyCalculatedDate.month)[1]
+            shareExpiryDate = datetime(quarterlyCalculatedDate.year, quarterlyCalculatedDate.month, monthEndDay)
+        elif expirySetting == Expiration.Monthly.value:
+            if last_week_start_date <= currentDate <= last_week_end_date:
+                # If in the last week of the month extend "expirationPeriod" number of months
                 monthlyCalculatedDate = currentDate + relativedelta(months=expirationPeriod)
+            else:
+                # If the user is not in the last week, consider this month into the calculation - extend for "expirationPeriod" - 1
+                monthlyCalculatedDate = currentDate + relativedelta(months=expirationPeriod - 1)
             monthEndDay = calendar.monthrange(monthlyCalculatedDate.year, monthlyCalculatedDate.month)[1]
             shareExpiryDate = datetime(monthlyCalculatedDate.year, monthlyCalculatedDate.month, monthEndDay)
         else:
             shareExpiryDate = None
 
         return shareExpiryDate
 
+    @staticmethod
+    def get_last_week_date_range(year, month):
+        last_day_of_month = calendar.monthrange(year, month)[1]
+        last_date = date(year, month, last_day_of_month)
+        last_week_day = last_date.weekday()
+        start_date = last_date - timedelta(days=last_week_day)
+
```

Comment:
Let me confirm what this function returns: start_date should be the last week_start date, it is calculated by substracting weekdays from the end_date of the month. the thing is that weekdays returns 0-6 (0 =Monday, 1=Tuesday.... 6=Sunday) if the end_date is a Monday, is it going to substract anything at all?

Reply:
And it won't fail as we are using <= in the comparison. I think it is good, I was just a bit confused. End of week is end of week (Sunday for most of the people) so I think it is good.